File: examples/python/analyze.py
#!/usr/bin/env python
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import List

import numpy as np
import onnxruntime as ort


def load_resources(res_dir: Path):
    tag2id = json.loads((res_dir / "tagset.json").read_text(encoding="utf-8"))
    word2id = json.loads((res_dir / "word_vocab.json").read_text(encoding="utf-8"))
    char2id = json.loads((res_dir / "char_vocab.json").read_text(encoding="utf-8"))
    tag_id2str = [t for t, _ in sorted(tag2id.items(), key=lambda kv: kv[1])]
    lex_path = res_dir / "lemma_lexicon.json"
    lemma_lex = json.loads(lex_path.read_text(encoding="utf-8")) if lex_path.exists() else {}
    return tag_id2str, word2id, char2id, lemma_lex


def encode(tokens: List[str], word2id, char2id, max_chars: int = 24):
    t = len(tokens)
    word_ids = np.zeros((1, t), dtype=np.int64)
    char_ids = np.zeros((1, t, max_chars), dtype=np.int64)
    for i, tok in enumerate(tokens):
        word_ids[0, i] = word2id.get(tok, 1)
        for j, ch in enumerate(tok[: max_chars - 1]):
            char_ids[0, i, j] = char2id.get(ch, 2)
    return word_ids, char_ids


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("tokens", nargs="*", help="space tokenized sentence")
    ap.add_argument("--model", default="artifacts/onnx/model.onnx")
    ap.add_argument(
        "--resources",
        default="rust/morphology_runtime/resources",
        help="directory with tagset/word_vocab/char_vocab",
    )
    ap.add_argument("--prefer-lexicon", action="store_true", help="prefer lemma_lexicon.json if present")
    args = ap.parse_args()

    res_dir = Path(args.resources)
    tag_id2str, word2id, char2id, lemma_lex = load_resources(res_dir)

    if not args.tokens:
        print("usage: onnx_analyze.py <token> [<token> ...]")
        return
    word_ids, char_ids = encode(args.tokens, word2id, char2id)

    sess = ort.InferenceSession(Path(args.model).as_posix(), providers=["CPUExecutionProvider"])
    outs = sess.run(None, {"word_ids": word_ids, "char_ids": char_ids})
    tag_logits, lemma_logits = outs
    tag_ids = tag_logits.argmax(axis=-1)[0]
    lemma_ids = lemma_logits.argmax(axis=-1)[0]

    # Convert lemma char IDs to strings until EOS=1
    id2ch = [ch for ch, _ in sorted(char2id.items(), key=lambda kv: kv[1])]
    results = []
    for i, tok in enumerate(args.tokens):
        tag = tag_id2str[tag_ids[i]]
        lemma_chars = []
        for cid in lemma_ids[i]:
            if cid == 1:
                break
            lemma_chars.append(id2ch[cid] if cid < len(id2ch) else "?")
        decoded = "".join(lemma_chars) or tok
        lemma = lemma_lex.get(tok, decoded) if args.prefer_lexicon else (decoded or lemma_lex.get(tok, decoded))
        results.append((tok, tag, lemma))

    for tok, tag, lemma in results:
        print(f"{tok}\t{tag}\t{lemma}")


if __name__ == "__main__":
    main()

File: examples/nodejs/analyze.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import ort from 'onnxruntime-node';

function loadJson(p) { return JSON.parse(fs.readFileSync(p, 'utf8')); }

function encode(tokens, word2id, char2id, maxChars=24) {
  const t = tokens.length;
  const wordIds = new BigInt64Array(1 * t);
  const charIds = new BigInt64Array(1 * t * maxChars);
  for (let i = 0; i < t; i++) {
    wordIds[i] = BigInt(word2id[tokens[i]] ?? 1);
    for (let j = 0; j < Math.min(tokens[i].length, maxChars - 1); j++) {
      const ch = tokens[i][j];
      charIds[i * maxChars + j] = BigInt(char2id[ch] ?? 2);
    }
  }
  return { wordIds, charIds, t, maxChars };
}

async function main() {
  const argv = process.argv.slice(2);
  let modelPath = 'artifacts/onnx/model.onnx';
  let resDir = 'data/processed';
  let preferLex = false;
  const args = [];
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--model' && argv[i+1]) { modelPath = argv[i+1]; i++; continue; }
    if (argv[i] === '--resources' && argv[i+1]) { resDir = argv[i+1]; i++; continue; }
    if (argv[i] === '--prefer-lexicon') { preferLex = true; continue; }
    args.push(argv[i]);
  }
  if (args.length === 0) {
    console.error('usage: node examples/nodejs/analyze.mjs [--model path] [--resources dir] <token> [<token> ...]');
    process.exit(1);
  }
  if (!fs.existsSync(path.join(resDir, 'tagset.json'))) {
    const alt = path.join(process.cwd(), '../../data/processed');
    if (fs.existsSync(path.join(alt, 'tagset.json'))) resDir = alt;
  }
  const tag2id = loadJson(path.join(resDir, 'tagset.json'));
  const word2id = loadJson(path.join(resDir, 'word_vocab.json'));
  const char2id = loadJson(path.join(resDir, 'char_vocab.json'));
  const id2tag = Object.entries(tag2id).sort((a,b)=>a[1]-b[1]).map(([k])=>k);
  const id2ch = Object.entries(char2id).sort((a,b)=>a[1]-b[1]).map(([k])=>k);
  let lemmaLex = {};
  const lexPath = path.join(resDir, 'lemma_lexicon.json');
  if (fs.existsSync(lexPath)) lemmaLex = loadJson(lexPath);

  const { wordIds, charIds, t, maxChars } = encode(args, word2id, char2id);

  if (!fs.existsSync(modelPath)) {
    const altModel = path.join(process.cwd(), '../../artifacts/onnx/model.onnx');
    if (fs.existsSync(altModel)) modelPath = altModel;
  }
  const session = await ort.InferenceSession.create(modelPath);
  const feeds = {
    word_ids: new ort.Tensor('int64', wordIds, [1, t]),
    char_ids: new ort.Tensor('int64', charIds, [1, t, maxChars])
  };
  const results = await session.run(feeds);
  const tagLogits = results['tag_logits'].data;
  const lemmaLogits = results['lemma_logits'].data;

  // Helper to argmax along last dim
  function argmaxRow(row, lastDim) {
    const out = new Int32Array(row.length / lastDim);
    for (let i = 0; i < out.length; i++) {
      let maxIdx = 0, maxVal = -Infinity;
      for (let j = 0; j < lastDim; j++) {
        const v = row[i*lastDim + j];
        if (v > maxVal) { maxVal = v; maxIdx = j; }
      }
      out[i] = maxIdx;
    }
    return out;
  }

  // tag logits: [1, t, K]
  const K = tagLogits.length / t;
  const tagIdx = argmaxRow(tagLogits, K);

  // lemma logits: [1, t, L, C]
  const LxC = lemmaLogits.length / t;
  // Find C by using maxChars assumption used in export (lemma_len=24 default). We'll infer L and C by searching a factor near 24.
  let L = 24; let C = Math.floor(LxC / L);
  const lemmaIdx = new Array(t);
  const preferLex = preferLex || process.env.PREFER_LEXICON === '1' || process.env.PREFER_LEXICON === 'true';
  for (let i = 0; i < t; i++) {
    const slice = lemmaLogits.subarray(i*L*C, (i+1)*L*C);
    lemmaIdx[i] = argmaxRow(slice, C);
  }

  for (let i = 0; i < t; i++) {
    const tag = id2tag[tagIdx[i]];
    const chars = [];
    for (const cid of lemmaIdx[i]) {
      if (cid === 1) break; // EOS
      chars.push(id2ch[cid] ?? '?');
    }
    const decoded = chars.length ? chars.join('') : args[i];
    const lemma = preferLex ? (lemmaLex[args[i]] ?? decoded) : (decoded || lemmaLex[args[i]] || decoded);
    console.log(`${args[i]}\t${tag}\t${lemma}`);
  }
}

main().catch(e => { console.error(e); process.exit(1); });
